Типизация 1

anufrp opened the pull request:
Компоненты + функции

anufrp committed:
Типизация 1

Компоненты + функции

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации

* нужно все файлы в `ts` переделать кроме редакса. Скрин https://disk.yandex.ru/i/W8EyUQ7JkyhTXQ . Все файлы в `pages` тоже. И в `app`

anufrp committed:
Типизация 2

Перевел в TS страницы в Pages, функции API

gennady-bars commented on src/utils/check-response.ts:
```diff
@@ -1,4 +1,4 @@
-export function checkResponse(res) {
+export const checkResponse = (res: Response) => {
```
Отлично, что знаете про тип `Response` для ответа от сервера

gennady-bars commented on src/services/hooks/useForm.ts:
```diff
@@ -0,0 +1,13 @@
+export function useForm(inputValues: {}) {
+    const [values, setValues] = React.useState<any>(inputValues);
+  
+    const handleChange = (event: SyntheticEvent) => {
```
Все события обработчиков нужно **правильно** типизировать. Это легко сделать. Для `onChange` есть тип `ChangeEvent<HTMLInputElement>`, а для `onSubmit` есть `FormEvent<HTMLFormElement>` в библиотеке `react`. Для обычных функций есть тип `SyntheticEvent` для их событий. Тип `Response` для ответа от сервера, тип `KeyboardEvent` для событий клавиатуры, тип `RouteProps` для рутов

Есть небольшая хитрость: можно всегда проверить, какой тип у функции или события. Для этого нужно навести мышкой на атрибут `onChange`, `onSubmit`, `onClick` и тд. Там высветится типизация того, что принимают эти атрибуты. 



```js

    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
      const {value, name} = event.target;

```

и не нужно будет вручную типизировать ` event.target  as HTMLInputElement;`

gennady-bars commented on src/pages/forgot-password/forgot-password.tsx:
```diff
@@ -22,13 +22,13 @@ export default function ForgotPasswordPage() {
     const {values, handleChange, setValues} = useForm({ email: '' });
 
-    const formSubmit = (e) => {
+    const formSubmit = (e: SyntheticEvent) => {
```
* Все события обработчиков нужно **правильно** типизировать. Это легко сделать. Для `onChange` есть тип `ChangeEvent<HTMLInputElement>`, а для `onSubmit` есть `FormEvent<HTMLFormElement>` в библиотеке `react`. Для обычных функций есть тип `SyntheticEvent` для их событий. Тип `Response` для ответа от сервера, тип `KeyboardEvent` для событий клавиатуры, тип `RouteProps` для рутов
* Есть небольшая хитрость: можно всегда проверить, какой тип у функции или события. Для этого нужно навести мышкой на атрибут `onChange`, `onSubmit`, `onClick` и тд. Там высветится типизация того, что принимают эти атрибуты. 




```js

    const formSubmit = (e: FormEvent<HTMLFormElement>) => {

```

gennady-bars commented on src/components/protected-route.tsx:
```diff
@@ -1,16 +1,28 @@
 import { getProfile } from '../services/actions/user';
 
+type TProtectRouteProps = {
+  onlyForAuth: boolean,
```
`ProtectedRoute` можно типизировать с помощью типа `RouteProps` из библиотеки `'react-router-dom'`.


Отличие только в `onlyForAuth`, значит, вот так нужно сделать


```js

import { Redirect, Route, RouteProps, useLocation } from 'react-router-dom';

type TProtectRouteProps = {
  onlyForAuth: boolean,
} & RouteProps 


```

gennady-bars commented on src/components/modal/modal.tsx:
```diff
@@ -1,18 +1,23 @@
-    const esc = useCallback((event) => {
+    const body = document.querySelector("body") as HTMLBodyElement;
+
+    const esc = useCallback((event: any) => {
```
* Все события обработчиков нужно **правильно** типизировать. Это легко сделать. Для `onChange` есть тип `ChangeEvent<HTMLInputElement>`, а для `onSubmit` есть `FormEvent<HTMLFormElement>` в библиотеке `react`. Для обычных функций есть тип `SyntheticEvent` для их событий. Тип `Response` для ответа от сервера, тип `KeyboardEvent` для событий клавиатуры, тип `RouteProps` для рутов



```js

    const esc = useCallback((event: KeyboardEvent) => {


```

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная


* Отлично, что знаете про тип `Response` для ответа от сервера
* Хорошо типизировали компоненты

но есть некоторые недочеты: 

* `app.utils.js` нужно тоже типизировать в папке `app`
* Все события обработчиков нужно **правильно** типизировать. Это легко сделать. Для `onChange` есть тип `ChangeEvent<HTMLInputElement>`, а для `onSubmit` есть `FormEvent<HTMLFormElement>` в библиотеке `react`. Для обычных функций есть тип `SyntheticEvent` для их событий. Тип `Response` для ответа от сервера, тип `KeyboardEvent` для событий клавиатуры, тип `RouteProps` для рутов
* Есть небольшая хитрость: можно всегда проверить, какой тип у функции или события. Для этого нужно навести мышкой на атрибут `onChange`, `onSubmit`, `onClick` и тд. Там высветится типизация того, что принимают эти атрибуты. 


**Можно лучше**

* Лучше типизировать все компоненты с помощью типа `FC` (даже те, которые не принимают `props`)



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

anufrp committed:
Типизация 3

Все компоненты типизированы как FC, событиям назначены конкретные типы, а не синтетик, типизирован редьюсер в APP, который отвечает за активный таб в ингредиентах. Переписана типизация ProtectedRoute на RouteProps

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
